flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/RetryableAsyncLookupFunctionDelegatorTest.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.operators.join;
+
+import org.apache.flink.streaming.util.MockStreamingRuntimeContext;
+import org.apache.flink.streaming.util.retryable.RetryPredicates;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.functions.AsyncLookupFunction;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.runtime.operators.join.lookup.ResultRetryStrategy;
+import org.apache.flink.table.runtime.operators.join.lookup.RetryableAsyncLookupFunctionDelegator;
+import org.apache.flink.table.runtime.util.RowDataHarnessAssertor;
+import org.apache.flink.table.types.logical.LogicalType;
+
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.apache.flink.table.data.StringData.fromString;
+
+/** Harness tests for {@link RetryableAsyncLookupFunctionDelegator}. */
+public class RetryableAsyncLookupFunctionDelegatorTest {
+
+    private final AsyncLookupFunction userLookupFunc = new TestingAsyncLookupFunction();
+
+    private final ResultRetryStrategy retryStrategy =
+            ResultRetryStrategy.fixedDelayRetry(3, 10, RetryPredicates.EMPTY_RESULT_PREDICATE);
+
+    private final RetryableAsyncLookupFunctionDelegator delegator =
+            new RetryableAsyncLookupFunctionDelegator(userLookupFunc, retryStrategy);
+
+    private static final Map<RowData, Collection<RowData>> data = new HashMap<>();
+
+    static {
+        data.put(
+                GenericRowData.of(1),
+                Collections.singletonList(GenericRowData.of(1, fromString("Julian"))));
+        data.put(
+                GenericRowData.of(3),
+                Arrays.asList(
+                        GenericRowData.of(3, fromString("Jark")),
+                        GenericRowData.of(3, fromString("Jackson"))));
+        data.put(
+                GenericRowData.of(4),
+                Collections.singletonList(GenericRowData.of(4, fromString("Fabian"))));
+    }
+
+    private final RowDataHarnessAssertor assertor =
+            new RowDataHarnessAssertor(
+                    new LogicalType[] {
+                        DataTypes.INT().getLogicalType(), DataTypes.STRING().getLogicalType()
+                    });
+
+    @Test
+    public void testLookupWithRetry() throws Exception {
```

Comment:
could you add more tests to cover the branches:
1. retryEnabled is disabled
2. different ResultRetryStrategy
3. ...

Reply:
Will add tests cover the case retry disabled, and also a custom result retry strategy using exponential backoff delay (though we only support fixed delay in sql for now)